policy: correct (lower) the dust threshold for Taproot outputs

darosior opened the pull request:
It was calculated as if it needed a compressed pubkey + a DER encoded
ECDSA signature.

Eventually i believe it would be good to prevent future overlooks by not applying the dust check to unknown witness versions and aborting if `witness_version > current_version`. Had a commit that starts doing that but removed it as it made this simple fix PR needlessly more complex.

sipa commented on src/policy/policy.cpp:
```diff
@@ -37,7 +37,13 @@ CAmount GetDustThreshold(const CTxOut& txout, const CFeeRate& dustRelayFeeIn)
+            nSize += (32 + 4 + 1 + (107 / WITNESS_SCALE_FACTOR) + 4);
+        } else {
+            // Taproot key spend, a single BIP341 signature
+            nSize += (32 + 4 + 1 + (65 / WITNESS_SCALE_FACTOR) + 4);
```
The outcome is the same, but I think the witness size is 66 (1 byte 0x01 to indicate the stack size, 1 byte 0x40 to indicate the size of the first element, 64 byte sig).

darosior replied:
Thanks, forgot the stack size byte

theStack commented:
Concept ACK

sipa commented:
Concept ACK

MarcoFalke commented on src/test/transaction_tests.cpp:
```diff
@@ -955,6 +955,15 @@ BOOST_AUTO_TEST_CASE(test_IsStandard)
     fIsBareMultisigStd = DEFAULT_PERMIT_BAREMULTISIG;
+
+    // Check Taproot outputs dust threshold
+    t.vout[0].scriptPubKey = CScript() << OP_1 << ParseHex("72ad3fb702bf1a2111b09c2bf1e72f4f52d4ceb2acdcfcd0c63abf70a59c36d6");
```
You can just use a repeating hex character to allow for better compression. 

```suggestion
    t.vout[0].scriptPubKey = CScript() << OP_1 << ParseHex("ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
```

Also, would be nice to add a test for OP_16 (as you change this too in this pull) and OP_0, because it is missing.

darosior replied:
Used repeating characters, tested for Segwit v0, and added test for future witness versions.

Awkward to test something that is almost certainly wrong, comforts me in the idea that we should refactor `GetDustThreshold` to loudly fail rather than returning an invalid value.

darosior committed:
policy: unit test Segwit v0 dust thresholds

Signed-off-by: Antoine Poinsot <darosior@protonmail.com>

darosior committed:
policy: correct (lower) the dust threshold for Taproot outputs

It was calculated as if it needed a compressed pubkey + a DER encoded
ECDSA signature.

Note that this commit intentionally keeps the existing behaviour of
applying the latest witness program version computation to all newer
versions.

Signed-off-by: Antoine Poinsot <darosior@protonmail.com>

Crypt-iQ commented on src/policy/policy.cpp:
```diff
@@ -37,7 +37,13 @@ CAmount GetDustThreshold(const CTxOut& txout, const CFeeRate& dustRelayFeeIn)
+            nSize += (32 + 4 + 1 + (107 / WITNESS_SCALE_FACTOR) + 4);
+        } else {
+            // Taproot key spend, a single BIP341 signature
+            nSize += (32 + 4 + 1 + (66 / WITNESS_SCALE_FACTOR) + 4);
```
More of a meta-comment:

If `GetDustThreshold` attempts to calculate the cost at which it's economical to spend an output, why is `nSize` initialized to `GetSerializedSize(txout)`? When spending the dust output, the size of the txout isn't in the spending tx unless the spending tx also spends to the same output type. This heuristic/initialization doesn't make sense to me because if we take P2WSH as an example (w/ dust limit of 330), `330 - fees < 330` it is impossible to spend to the same output type unless other inputs are added. It seems to me that if witness version non-zero is being changed, this could be a place to fix or clarify this calculation for witness version non-zero.

ariard commented:
Concept ACK

> Eventually i believe it would be good to prevent future overlooks by not applying the dust check to unknown witness versions and aborting if witness_version > current_version.

IIUC, you're aiming to return `IsDust()` == true for any transaction with witness outputs superior to current version, wouldn't this be a revert of https://github.com/bitcoin/bitcoin/pull/15846 ? As we don't know what would be the minimal spent or we can pick-up an arbitrary amount e.g 1000 sats, and consider the effective dust threshold for future type will be lower. And as such minimize risks of policy _tightening_ changes.

darosior commented:
No, i'm aiming to do the opposite: not call IsDust for witness versions >1, effectively having IsDust == false for any witness version from the future. Since, as you mention it, we can't know the size in advance.
‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
Le mardi 24 août 2021 à 11:21 PM, Antoine Riard ***@***.***> a écrit :

> Concept ACK
>
>> Eventually i believe it would be good to prevent future overlooks by not applying the dust check to unknown witness versions and aborting if witness_version > current_version.
>
> IIUC, you're aiming to return IsDust() == true for any transaction with witness outputs superior to current version, wouldn't this be a revert of [#15846](https://github.com/bitcoin/bitcoin/pull/15846) ? As we don't know what would be the minimal spent or we can pick-up an arbitrary amount e.g 1000 sats, and consider the effective dust threshold for future type will be lower. And as such minimize risks of policy tightening changes.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, [view it on GitHub](https://github.com/bitcoin/bitcoin/pull/22779#issuecomment-904985220), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AFLK3F233XM5Z52Y2EHXUU3T6QEL7ANCNFSM5CUWHKOQ).
> Triage notifications on the go with GitHub Mobile for [iOS](https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675) or [Android](https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email).

ariard commented on src/policy/policy.cpp:
```diff
@@ -37,7 +37,13 @@ CAmount GetDustThreshold(const CTxOut& txout, const CFeeRate& dustRelayFeeIn)
+            // P2WPKH spend, a compressed public key + a DER-encoded ECDSA signature
+            nSize += (32 + 4 + 1 + (107 / WITNESS_SCALE_FACTOR) + 4);
+        } else {
+            // Taproot key spend, a single BIP341 signature
```
nit: If you intent to refer to the signature standard, should say BIP340 no? BIP341 describe the consensus rules applying the schnorr signature scheme.

darosior replied:
Right, thanks. Doc-off-by-one!

ariard commented:
> No, i'm aiming to do the opposite: not call IsDust for witness versions >1, effectively having IsDust == false for any witness version from the future.

Hmmmm so you can propagate transactions SegWit v2+ with `nValue` == 0 and as such freely write on the UTXO set? Unknown witness outputs are standard iirc.

darosior commented:
> Hmmmm so you could propagate transactions SegWit v2+ with nValue == 0 and as such freely write on the UTXO set? Unknown witness outputs are standard iirc.

You are right, i shouldn't reply to Github comments at 11:21PM :p
Maybe it could be made of a fixed value (this one, or another large one so we don't tighten it when introducing a new version). Anyways i don't intend to (address it in this PR and) modify the existing behaviour, just to refactor the code to prevent a future overlook.

TheBlueMatt commented:
Concept NACK. I don't think it makes sense to be lowering the dust levels in bitcoin generally. The context for the dust levels was calculated many years ago, I don't think it applies anymore.

benthecarman reviewed (approved):
Code Review ACK 9155f6cf738c790c389e105ce48d895634dd084f

JeremyRubin commented on src/policy/policy.cpp:
```diff
@@ -37,7 +37,13 @@ CAmount GetDustThreshold(const CTxOut& txout, const CFeeRate& dustRelayFeeIn)
         // sum the sizes of the parts of a transaction input
         // with 75% segwit discount applied to the script size.
-        nSize += (32 + 4 + 1 + (107 / WITNESS_SCALE_FACTOR) + 4);
+        if (witnessversion == 0) {
```
prefer a switch statement here

JeremyRubin commented:
Concept NACK --

I think we should either remove the dust limit entirely or not muck around with it.

Dust limits are not incentive compatible with miner preferences and if they are they can be configured by users individually.

That said, it's CR-Ack for me for this code, which seems to be improving the dust nSize calculation (which should be independent of any actual fee policy, the code could be cleaned up to make it have a min somewhere later but we shouldn't confuse byte counting with thresholds).

MarcoFalke commented:
Maybe split off the added tests from the behaviour change, because they seem useful on their own?

darosior commented:
@MarcoFalke done in #22846

@TheBlueMatt i too am of the opinion that we should not decrease the dust thresholds, but i think that if we want to keep the same dust thresholds for Taproot outputs we should instead raise `DUST_RELAY_TX_FEE` instead of keeping the wrong computation. But this would unfortunately be a security issue for (some) L2s relying on it..

I don't think the rationale of "dust threshold was calculated many years ago, let's not fix it for a new output type unused on the network yet" holds. It was adapted for Segwit v0 and now Segwit v1 should be "penalized"?
In any case, i think that even if we want to keep it larger for Taproot outputs than for Segwit v0 outputs we should  document it instead of having the current "at least a 33-bytes pubkey and an ECDSA signature are required to spend" rationale for any Segwit output.

TheBlueMatt commented:
> @TheBlueMatt i too am of the opinion that we should not decrease the dust thresholds, but i think that if we want to keep the same dust thresholds for Taproot outputs we should instead raise DUST_RELAY_TX_FEE instead of keeping the wrong computation. But this would unfortunately be a security issue for (some) L2s relying on it..

Indeed, I don't think we can in any practical sense increase the dust threshold for deployed things, for exactly the reason you specify. I also agree with you that, ideally, we'd increase `DUST_RELAY_TX_FEE`, but given that we agree that we cannot, the only thing we can do is not make the problem worse for, which I'd argue this does.

I'm not really sure why "consistency" here matters, the numbers are too low, reducing them is not great, so lets not do that :).

darosior commented:
I think Matt is right. Closing this one in favour of documenting the behaviour in https://github.com/bitcoin/bitcoin/pull/22863 .

luke-jr commented:
> But this would unfortunately be a security issue for (some) L2s relying on it..

Then those L2s are broken. Policy is never a guarantee and should not be treated like one.

IMO you should reopen this.
